### PR TITLE
Add an IPython widget

### PR DESCRIPTION
**Pull request by blink1073**
Makes the QGrid into an IPython Widget (valid for IPython 3 and 4). 
- Adds the ability to edit contents with validation for numerical types.
- Categorical types automatically create a dropdown list.
- Date types create a date editor.
- Can add/remove rows.
- Can export the grid for a static view.

qgrid_demo5

Categorical data:

screen shot 2015-08-09 at 10 58 24 am

### CODE REVIEW

**Comment by blink1073** [comment]
@ssanderson, I think it makes sense to keep editing and displaying separate.  If the user wants to embed the document, they can call `show_grid` when they are done.

**Comment by ssanderson** [comment]
@blink1073 thanks for this.  I'm not going to have bandwidth to look at this seriously for a couple days. cc @TimShawver in case you think you can get to this in a more timely manner.

**Comment by blink1073** [comment]
FYI, I'm still working on making this more robust to input, and resizable.

**Comment by ssanderson** [comment]
@blink1073 ok.  Let us know when we should take a look.

File: qgrid/grid.py [hunk 1]
```diff
@@ -81,3 +85,108 @@ def _ipython_display_(self):
...
+            self._cdn_base_url = "/nbextensions/qgridjs"
+
+    def add_row(self, value=None):
+        """Append a row at the end of the dataframe."""
+        df = self.df
+        if not df.index.is_integer():
+            msg = 'alert("Cannot add a row a table with a non-integer index")'
+            display(Javascript(msg))
+            return
+        last = df.loc[df.index[-1], :]
```
Comment (ssanderson) [inline qgrid/grid.py]: Noticed while skimming:

I think this is just `df.iloc[-1]`?

Comment (blink1073) [inline qgrid/grid.py]: Ha!, perhaps, I confess I do not use pandas regularly, so any help in that regard is appreciated.

**Comment by blink1073** [comment]
@ssanderson, I'm finished.  Added:

- resizing and handling of the scroll bar when adding/removing rows
- widget installation method recommended by the ipywidgets developers
- an `auto_edit` parameter

**Comment by blink1073** [comment]
Added support for Categorical data:

screen shot 2015-08-09 at 10 58 24 am

File: qgrid/grid.py [hunk 2]
```diff
@@ -81,3 +89,122 @@ def _ipython_display_(self):
...
+        if self.remote_js:
+            self._cdn_base_url = REMOTE_URL
+        else:
+            self._cdn_base_url = "/nbextensions/qgridjs"
+
+    def add_row(self, value=None):
+        """Append a row at the end of the dataframe."""
+        df = self.df
+        if not df.index.is_integer():
+            msg = 'alert("Cannot add a row a table with a non-integer index")'
```
Comment (TimShawver) [inline qgrid/grid.py]: You're just missing the word "to".  Cannot add a row **to** a table with a non-integer index.

**Comment by TimShawver** [comment]
@blink1073 I've been playing with this locally today and it looks great.

My initial reaction is basically the same as @ssanderson's: I'm tempted to just drop the old implementation and always render qgrid with your new widget-based implementation.  As it stands there's a bunch of code duplicated between the two implementations so it would be nice to consolidate.  

I think adding an **Embed** option like @ssanderson suggested seems like the best of both worlds, but if we can't figure out how to do that then I'd be fine with keeping both implementations as you've proposed.

I probably won't have time to finish testing and get this merged until next weekend, at which time I can also look into this question of whether we can embed the widget output into a static view of the notebook (unless you're able to look into it sooner).

**Comment by blink1073** [comment]
It looks like `display(*widgets)` suppresses any other html outputs, so I don't see an `Export` button working:

```python
from IPython.html.widgets import Button
from IPython.display import HTML, display
b = Button(description='Say Hello')
def cb(value):
    HTML('<h1>hello</h1>')
    print('hello')
    
b.on_click(cb)
display(b)
```

Clicking the button does not produces the HTML output:

screen shot 2015-08-10 at 6 06 43 pm

**Comment by blink1073** [comment]
I'm going to take another look at `Embed`...

**Comment by blink1073** [comment]
And here it is.  Next step is to streamline the process to always use the widget, and make Embed use the current dataframe.

qgrid_demo4

**Comment by TimShawver** [comment]
Wow, awesome.  Haven't tried it locally yet but the code makes sense to me, and the behavior looks right in the gif.

**Comment by TimShawver** [comment]
@blink1073 are you still working on this? If you want help with refactoring the show_grid function to use the widget implementation, just let me know.

**Comment by blink1073** [comment]
I should be able to knock out the refactor today.

**Comment by TimShawver** [comment]
Nice :sunglasses:

**Comment by blink1073** [comment]
Okay, done.

**Comment by blink1073** [comment]
Darn, this breaks the remote viewing capability: see https://gist.github.com/blink1073/ad3e65953e532547f623

**Comment by blink1073** [comment]
Fixed: http://nbviewer.ipython.org/gist/blink1073/e189ea1217199cdb4cdc

**Comment by TimShawver** [comment]
Awesome, looks great.  It worked well on IPython 3.2.0 and Python 2 but was seeing some issues on other versions.  I haven't done much trouble shooting so it could be something wrong with my environment, but I'm including the stack traces below in case you can figure them out.  I have more testing to do so I'll keep you posted on other things I find.

What version of IPython and Python were you using while developing?

I get this stack trace when I run `import qgrid` on IPython 4.0.0 and Python 3:
```
import qgrid
/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/html.py:14: ShimWarning: The `IPython.html` package has been deprecated. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.
  "`IPython.html.widgets` has moved to `ipywidgets`.", ShimWarning)
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/utils/shimmodule.py in __getattr__(self, key)
     89         try:
---> 90             return import_item(name)
     91         except ImportError:

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/utils/importstring.py in import_item(name)
     30         package, obj = parts
---> 31         module = __import__(package, fromlist=[obj])
     32         try:

ImportError: No module named 'ipywidgets'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-1-a4f5c542216e> in <module>()
----> 1 import qgrid

/Users/tshawver/repo/qgrid-blink1073/qgrid/__init__.py in <module>()
----> 1 from .grid import (
      2     set_defaults,
      3     set_grid_option,
      4     show_grid
      5 )

/Users/tshawver/repo/qgrid-blink1073/qgrid/grid.py in <module>()
    163 
    164 
--> 165 class QGridWidget(widgets.DOMWidget):
    166     _view_name = Unicode('QGridView', sync=True)
    167     _df_json = Unicode('', sync=True)

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/utils/shimmodule.py in __getattr__(self, key)
     90             return import_item(name)
     91         except ImportError:
---> 92             raise AttributeError(key)

AttributeError: DOMWidget
```
I get this stack trace when I run show_grid on IPython 3.2.1 and Python 3:
```
qgrid.show_grid(spy, remote_js=False)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-a7a1de645982> in <module>()
----> 1 qgrid.show_grid(spy, remote_js=False)

/Users/tshawver/repo/qgrid-blink1073/qgrid/grid.py in show_grid(data_frame, remote_js, precision, grid_options)
    149     grid = QGridWidget(df=data_frame, precision=precision,
    150                        grid_options=json.dumps(grid_options),
--> 151                        remote_js=remote_js)
    152 
    153     add_row = widgets.Button(description="Add Row")

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/html/widgets/widget.py in __init__(self, *pargs, **kwargs)
    479 
    480     def __init__(self, *pargs, **kwargs):
--> 481         super(DOMWidget, self).__init__(*pargs, **kwargs)
    482 
    483         def _validate_border(name, old, new):

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/html/widgets/widget.py in __init__(self, **kwargs)
    151         """Public constructor"""
    152         self._model_id = kwargs.pop('model_id', None)
--> 153         super(Widget, self).__init__(**kwargs)
    154 
    155         Widget._call_widget_constructed(self)

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/config/configurable.py in __init__(self, **kwargs)
     84         # This should go second so individual keyword arguments override
     85         # the values in config.
---> 86         super(Configurable, self).__init__(**kwargs)
     87 
     88     #-------------------------------------------------------------------------

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/IPython/utils/traitlets.py in __init__(self, *args, **kw)
    563         # notifications.
    564         for key, value in iteritems(kw):
--> 565             setattr(self, key, value)
    566 
    567     def _notify_trait(self, name, old_value, new_value):

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/traitlets/traitlets.py in __set__(self, obj, value)
    448 
    449     def __set__(self, obj, value):
--> 450         new_value = self._validate(obj, value)
    451         try:
    452             old_value = obj._trait_values[self.name]

/Users/tshawver/.virtualenvs/ipython3/lib/python3.4/site-packages/traitlets/traitlets.py in _validate(self, obj, value)
    470         if hasattr(self, 'validate'):
    471             value = self.validate(obj, value)
--> 472         if obj._cross_validation_lock is False:
    473             value = self._cross_validate(obj, value)
    474         return value

AttributeError: 'QGridWidget' object has no attribute '_cross_validation_lock'
```

File: qgrid/qgridjs/qgrid.widget.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                msg = {'type': 'remove_row', 'row': cell.row, 'id': data.id};
+                this.updateSize();
+                this.send(msg);
+
+            } else if (msg.type === 'add_row') {
+                var dd = sgrid.getData();
+                dd.addItem(msg);
+                dd.refresh();
+                this.updateSize();
+                this.send(msg);
```
Comment (TimShawver) [inline qgrid/qgridjs/qgrid.widget.js]: You're missing a `}` to close the `else if`

Comment (blink1073) [inline qgrid/qgridjs/qgrid.widget.js]: Fixed, thanks.

**Comment by blink1073** [comment]
@TimShawver, I'm using IPython 3.2.1 with Python 3.4.3.  It looks like you had some stale state when going back to IPython 3, since it was looking for traitlets.
I'll upgrade an try on IPython 4.

**Comment by blink1073** [comment]
It works fine on IPython 4 for me as well, with no deprecation warnings.  It looks like you were missing the `ipywidgets` package.

**Comment by TimShawver** [comment]
Nice, I started with a fresh virtualenv and things are looking good so far.  Haven't tried IPython 4 yet, I'll do that again soon.

Another thing I noticed...I get the following error printed out under the grid when I try to remove a row:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/Users/tshawver/repo/qgrid-blink1073/qgrid/grid.py in _handle_qgrid_msg(self, widget, content, buffers)
    259             return
    260         if content['type'] == 'remove_row':
--> 261             self.df.drop(content['row'], inplace=True)
    262 
    263         elif content['type'] == 'cell_change':

/Users/tshawver/.virtualenvs/qgrid-widget3/lib/python3.4/site-packages/pandas/core/generic.py in drop(self, labels, axis, level, inplace, errors)
   1595                 new_axis = axis.drop(labels, level=level, errors=errors)
   1596             else:
-> 1597                 new_axis = axis.drop(labels, errors=errors)
   1598             dropped = self.reindex(**{axis_name: new_axis})
   1599             try:

/Users/tshawver/.virtualenvs/qgrid-widget3/lib/python3.4/site-packages/pandas/core/index.py in drop(self, labels, errors)
   2568         if mask.any():
   2569             if errors != 'ignore':
-> 2570                 raise ValueError('labels %s not contained in axis' % labels[mask])
   2571             indexer = indexer[~mask]
   2572         return self.delete(indexer)

ValueError: labels [4] not contained in axis
```
The grid I tested on was the spy data frame from the demo notebook, which doesn't have an integer index (not sure if that matters, just pointing it out since adding was disabled for that reason).

**Comment by blink1073** [comment]
Hmm, I tried it with the same dataset and didn't get an error.  I'm using pandas 0.16.2.

**Comment by TimShawver** [comment]
@blink1073 sorry about the delay, just got back to looking at this.  I'm still getting that stack trace when I try to remove a row, but in the process of trying to fix it I came up with a couple of new suggestions:

##### 1. Switch to using latest methods for loading widget JS
I was reading about widget best practices and learned that there's a new strategy for distributing the JS that's associated with a widget, as we need to do here.  Take a look at this paragraph, taken from https://ipython.org/ipython-doc/3/whatsnew/version3_widget_migration.html:

> If you distribute your custom widget, you may be using display and Javascript to publish the widget’s Javascript to the front-end. That is no longer the recommended way of distributing widget Javascript. Instead have the user install the Javascript to his/her nbextension directory or their profile’s static directory. Then use the new _view_module and _model_module traitlets in combination with _view_name and _model_name to instruct require.js on how to load the widget’s Javascript. The Javascript is then loaded when the widget is used for the first time.

I also came across what I think is a good example of this strategy being put to use: https://github.com/takluyver/mobilechelonian

##### 2. Make buttons hidden by default, to reduce UI clutter
I think the widget looks slightly less cool with the Add Row / Remove Row / Export buttons on top of the grid.  I'm thinking ideally we'd make those buttons optional and default to being hidden (i.e. we'd add a new `show_toolbar` option that could be set with the `set_defaults` function, and would default to `False`).

I think I'd be fine merging without 2, but the 1 seems like it's worth doing.  Let me know if you'd like me to help out with either of these changes (although I might not have time for a couple of weeks due an upcoming vacation).

**Comment by blink1073** [comment]
Hi @TimShawver, I addressed 2), but could not quite get 1) to work.  The custom notebook example that ships with the latest `ipywidgets` still uses the method I am [using](https://github.com/ipython/ipywidgets/blob/master/examples/Custom%20Widget%20-%20Hello%20World.ipynb), and the mechanism for loading widgets still hasn't quite settled down, so I think this is fine as is.

**Comment by TimShawver** [comment]
@blink1073 awesome, thanks for doing that.  Regarding 1) I just pushed a branch which seems to be working fine using the new JS loading method.  Here's a PR: https://github.com/quantopian/qgrid/pull/29

If you have time maybe you could try it out locally and let me know if it works for you.

**Comment by TimShawver** [comment]
Merged in this PR https://github.com/quantopian/qgrid/pull/29

**Pull request closed without merging**